openedx/core/djangoapps/discussions/docs/decisions/0002-discussion-plugin-api.rst

```diff
@@ -0,0 +1,71 @@
+Discussions Plugin API
+======================
+
+Status
+------
+
+Proposal
+
+Context
+-------
+
+Currently the Open edX platform provides its own discussions experience,
+using a form tool that ships with the platform and is well integrated
+into it. There is a common need to support other, external discussion
+tools into the platform.
+
+To make it possible for the platform to support integration with other
+discussion tools, this ADR proposes a system for a new type of plugin
+that allows extending the platform by adding other discussion tools
+that can integrate more closely with the platform.
+
+
+Decision
+--------
+
+We can enable a new type of plugin, a discussion app, that will provide
```

Comment:
Question about the internal python APIs - do we consider these to be an optional integration method from the standard APIs (which I'm assuming mean REST APIs?). Does it exist because we need it for our own home grown forums plugin?  The rest of this is predicated on that assumption, so I may be off base here:

My expectation is that third-party discussions tools like Yellowdig, Piazza, Discourse, etc., would likely not be able or willing to take advantage of the internal APIs, and would generally integrate with Open edX via standard APIs. If they wanted to use "internal" APIs, doing so would mean packaging their software or an "adapter" to their software as an Open edX-compatible django app, right?  Since internal django apps/plugins have historically been problematic in terms of sandboxing/maintaining clean boundaries, I'm wondering if we even want to have that option.  

If we're including the internal APIs for our own, homegrown forums app, might be better for us to re-work our forums app to use the same standard APIs that third-party discussions plugins are likely to use, and jettison the need for the internal APIs?  (I may be missing something here!)

And if we don't need those internal APIs, do we even need to model these as server-side django apps installed into edx-platform anymore?  Or do they become independent micro-services that communicate with the LMS via the standard APIs?

Reply:
I realize, now that I've written the above, that our goal for milestone 1 of this project is to create "Course Apps as Backend Plugin or Hard-coded"

So the above should be considered more academic/strategic and not a blocker for this PR. :)

Reply:
I believe it might be possible to create a plugin-less approach to support discussion tools that can be embedded via LTI, however the aim of having these APIs is to support deeper integration of discussion tools into the platform. 

In general, I think of Discussion apps are a kind of course app, so it does involve a backend plugin. For instance, we need some way to know how to embed in-context discussions, or how to map internal users to users in the external forum. LTI is one way to maintain an internal and external user mapping, but some discussion tools might not support that (for instance Discourse). 

The Piazza plugin currently just has some metadata about the plugin that can be shown to course authors, and has a view that embeds it in the discussion tab via LTI.

Having these become independent micro-services is definitely one way to go. i.e. a piazza service that bridges the piazza API and the edx-platform API, however that makes it more complex to deploy. There is definitely some need for a backend to handle operations that a frontend just can't, whether it's in plugin or IDA.